Add virtio-fs support (alternative to virtio-9p)

stefanha opened the pull request:
These patches add several configuration.toml options for the virtio-fs shared file system.  For more information about virtio-fs, see https://virtio-fs.gitlab.io/.

This pull request is a preview.  virtio-fs has not been merged in Linux or QEMU yet.

To use virtio-fs, set enable_virtio_fs true and set virtio_fs_daemon to the path of your passthrough_ll file system daemon (code available at https://gitlab.com/virtio-fs/libfuse).  A full howto based on Kata 1.4.0 and a Fedora initramfs is here: https://virtio-fs.gitlab.io/howto-kata.html.

Tweaking the cache size, cache mode, and enabling shared version metadata is mostly for development at this point.  The default settings are recommended.

grahamwhaley commented:
Marking as WIP/DNM to show the preview status.
I'm presuming you know already @stefanha that the govmm file change for QEMU additions will need to go via the [govmm repo](https://github.com/intel/govmm) and then get re-vendored back into here - so, you may want to post and progress that one separately (as it may need to land first).

devimc commented:
thanks @stefanha, could you please file an issue and add `fixes` in your commit message, for example 
`fixes #99999`

devimc commented:
@stefanha https://github.com/kata-containers/community/blob/master/CONTRIBUTING.md#patch-format

ydjainopensource commented:
@stefanha Any updates here?

jodh-intel commented:
I guess we're still blocked on upstream. But it might be a good idea to keep this PR branch conflict free so we're ready to jump on it when the time is right?

jodh-intel commented:
Hi @stefanha - could you replace the questions marks in the table below so we have some tracking information on virtio-fs support?

| Dependency | Status | URL |
|-|-|-|
| kernel | ? | ? |
| virtio spec | ? | ? |
| qemu | ? | ? |

stefanha commented:
@jodh-intel 
| Dependency | Status | URL |
| --- | --- | --- |
| kernel | ? | https://lkml.org/lkml/2018/12/10/615 |
| virtio spec | ? | (draft) https://github.com/stefanha/virtio/commit/e1cac3777ef03bc9c5c8ee91bcc6ba478272e6b6 |
| qemu | ? | https://patchwork.kernel.org/cover/10722021/ |
| libfuse | ? | ? |

Currently the VIRTIO spec changes are being worked on.  They are the dependency for everything else.

raravena80 commented:
@stefanha ping, any updates? Thx! This would be awesome to get in!

stefanha commented:
@raravena80 Work is ongoing but the status hasn't changed.

jodh-intel commented on virtcontainers/qemu.go:
```diff
@@ -1230,7 +1260,36 @@ func (q *qemu) addDevice(devInfo interface{}, devType deviceType) error {
 	case types.Volume:
-		q.qemuConfig.Devices = q.arch.append9PVolume(q.qemuConfig.Devices, v)
+		if q.config.VirtioFS {
+			q.Logger().Info("Adding vhost-user-fs-pci volume")
```
Could you tweak this message (and the 9p one below) to use a structure format to make it more parseable - something like:

```go
q.Logger().WithField("volume-type", "vhost-user-fs-pci").Info("adding volume")
   :
q.Logger().WithField("volume-type", "virtio-9p").Info("adding volume")
```

jodh-intel commented on vendor/github.com/intel/govmm/qemu/qemu.go:
```diff
@@ -92,6 +92,9 @@ const (
 	//VhostUserBlk represents a block vhostuser device type.
 	VhostUserBlk DeviceDriver = "vhost-user-blk-pci"
 
+	//VhostUserFS represents a virtio-fs vhostuser device type
```
Please get these changes into https://github.com/intel/govmm first, and then re-vendor the changes into the runtime (see https://github.com/kata-containers/community/blob/master/VENDORING.md).

jodh-intel commented on virtcontainers/device/drivers/vhost_user_fs.go:
```diff
@@ -0,0 +1,72 @@
+}
+
+func (device *VhostUserFSDevice) Detach(devReceiver api.DeviceReceiver) error {
+	skip, err := device.bumpAttachCount(true)
```
This looks wrong - the `true` parameter is for attach only, but this is a detach method...?

jodh-intel commented on pkg/katautils/config.go:
```diff
@@ -102,6 +102,11 @@ type hypervisor struct {
 	DefaultBridges          uint32 `toml:"default_bridges"`
 	Msize9p                 uint32 `toml:"msize_9p"`
 	DisableBlockDeviceUse   bool   `toml:"disable_block_device_use"`
+	VirtioFS                bool   `toml:"enable_virtio_fs"`
```
It would be useful to see something in the output of `kata-runtime kata-env` to show that virtio-fs is active. This saves users from having to find the correct configuration file (the runtime support multiple locations).

For examples, see #702, #937.

stefanha commented:
Thanks for the review, @jodh-intel.  I'll make the changes in the next revision.

amshinde commented on virtcontainers/qemu.go:
```diff
@@ -1230,7 +1249,33 @@ func (q *qemu) addDevice(devInfo interface{}, devType deviceType) error {
+			}
+			id := hex.EncodeToString(randBytes)
+
+			// TODO make this unique so multiple volumes can be added!
```
Isnt this unique since you are using a random string(id) as part of the path ?

amshinde commented on virtcontainers/qemu.go:
```diff
@@ -1230,7 +1249,33 @@ func (q *qemu) addDevice(devInfo interface{}, devType deviceType) error {
+
+			vhostDev := config.VhostUserDeviceAttrs{
+				Tag: v.MountTag,
+				Type: config.VhostUserFS,
```
Indent seems to be off here. @stefanha Can you format the files using `gofmt`?

stefanha replied:
Thanks, will fix in the next revision.

amshinde commented on cli/config/configuration-qemu.toml.in:
```diff
@@ -121,6 +121,9 @@ virtio_fs_cache_size = @DEFVIRTIOFSCACHESIZE@
 virtio_fs_cache = "@DEFVIRTIOFSCACHE@"
 
+# Use shared version metadata
+virtio_fs_shared_versions = @DEFVIRTIOFSSHAREDVERSIONS@
```
@stefanha Can you elaborate more on what this option does, pros and cons of using this?

stefanha replied:
This is an experimental feature and it's not ready for use.  I will remove it for the time being since it relies on an external daemon (ireg) that isn't managed by kata-runtime yet.  Once this feature is working it will make the cache=always mode coherent across shared volumes.

sboeuf replied:
Agreed. I think the first step is to get virtio-fs early supported with the things that are reliable in virtio-fs for now. We can definitely add the `ireg` support with a later follow up PR :)

amshinde commented:
@stefanha Can you take a look at the comments here? Although virtio-fs has not merged upstream, we are planning to add early support for virtio-fs. It would be great if you could submit a patch for VhostUserFs device to the `govmm` repo as well.

stefanha commented:
> @stefanha It would be great if you could submit a patch for VhostUserFs device to the `govmm` repo as well.

The VhostUserFs device has been merged into govmm:
https://github.com/intel/govmm/pull/86

sboeuf commented:
@jodh-intel @devimc @grahamwhaley @jcvenegas @bergwolf @WeiZhang555 Please take a look as we're trying to move forward with the early support of virtio-fs for Kata. Think about this PR as a high prio one since we don't want to have it pending for months :)

jodh-intel commented on virtcontainers/device/drivers/vhost_user_fs.go:
```diff
@@ -0,0 +1,72 @@
+
+	defer func() {
+		if err == nil {
+			device.AttachCount = 1
```
I'm rather confused by this: the function calls `device.bumpAttachCount(true)` to increase `AttachCount` and then explicitly sets `AttachCount = 1` if no error. I see the same logic in reverse in `Detach()`. It seems therefore that the only valid `AttachCount` values are `0` and `1` but can't that be checked for rather than overriding what `bumpAttachCount()` is doing?

Maybe a few comments explaining the intent would help here?

WeiZhang555 replied:
@jodh-intel I think I am the one to be blamed :crying_cat_face: 

The logic is confusing as you said but it's right. 
* `AttachCount == 0`: do real attach, and set AttachCound = 1;
* `AttachCount==1`: AttachCount++, but skip the real attact.
* `AttachCount==IntMax`: error

I'll create a new PR to improve the readability.

WeiZhang555 replied:
Here it is: https://github.com/kata-containers/runtime/issues/1392

It's assigned to me, I'll handle it today.

WeiZhang555 replied:
PR #1394 for this

devimc commented on virtcontainers/qemu.go:
```diff
@@ -573,6 +575,34 @@ func (q *qemu) startSandbox(timeout int) error {
+		if q.config.VirtioFSSharedVersions {
+			args = append(args, "-o", "shared")
+		}
+		args = append(args, "/")
```
Q. why `/` ? are you given access to the whole rootfs? can you please add a comment?

stefanha replied:
This is removed in the next revision of this PR.  The "/" was a dummy argument that has now been removed from virtiofsd.  The shared directory is specified by the `virtiofsd -o source path/to/dir` option.

WeiZhang555 commented on virtcontainers/qemu.go:
```diff
@@ -549,6 +551,34 @@ func (q *qemu) startSandbox() error {
 	}()
 
+	if q.config.VirtioFS {
+                var cmd *exec.Cmd
```
`gofmt` could help

WeiZhang555 commented:
Overall it looks good to me, nice work :smile: 

PR needs rebase, and I met some problem when testing the PR locally, the filesystem keeps reporting "Busy" even a `ls` command can't get a result though the container is up successfully, still trying to figure out why...

sboeuf commented on virtcontainers/qemu.go:
```diff
@@ -573,6 +575,23 @@ func (q *qemu) startSandbox(timeout int) error {
+		// of this child process.
+		sourcePath := filepath.Join(kataHostSharedDir, q.id)
+
+		cmd := exec.Command(q.config.VirtioFSDaemon, "-o", "virtio_socket=" + sockPath, "-o", "source=" + sourcePath, "/")
```
This part (spawning the `virtiofsd` binary) is not QEMU specific. It can be shared across hypervisors as long as they support `virtio-fs`. That's why it'd be better to have it implemented from hypervisor.go directly so that each hypervisor implementation could reuse the common code. Or, to invoke this from the generic `sandbox.go` code.

stefanha replied:
I looked into this but found sandbox.go and hypervisor.go don't offer convenient places to put this code.  The issue is that qemu.go creats the vm directory so sandbox.go cannot launch the daemon before calling hypervisor.startSandbox().  And putting this function into hypervisor.go is awkward because it relies on paths that come from qemu.go (or another caller who has access to objects).

Let's leave it in qemu.go until there is a need to move it out for Firecracker.

sboeuf commented on Makefile:
```diff
@@ -163,6 +163,8 @@ DEFENTROPYSOURCE := /dev/urandom
 DEFDISABLEBLOCK := false
 DEFVIRTIOFS := false
 DEFVIRTIOFSDAEMON :=
+# Default DAX mapping cache size in GiB
```
Maybe `MiB` would be more appropriate if we don't want a too large cache region by default.

stefanha replied:
Fixed in the next revision.

sboeuf commented:
Hi @stefanha, I was wondering if you had an idea when you will be able to rework this PR? I think everybody had the time to make a first pass of review.
And of course, please let us know if you don't have the bandwidth for that, someone from the community might be able to help finishing.

WeiZhang555 commented:
I think this should be an "experimental" feature since all necessary patches haven't been accepted by kernel/qemu upstream community.

Experimental feature support: https://github.com/kata-containers/runtime/pull/1227

I think it can turn from experimental to formal feature once all patches are mainstreamed in any Kata release and not have to be 2.0, since it's a new feature. We can change it to default in release 2.0 though.

egernst commented:
@WeiZhang555 agreed.

stefanha commented:
@sboeuf I've rebased onto Kata 1.6.0 and updated the govmm vendoring.  Now I will work through the review feedback and retest.

sboeuf commented:
Thanks for the heads up @stefanha

stefanha commented on cli/config/configuration-qemu.toml.in:
```diff
@@ -97,6 +97,9 @@ default_memory = @DEFMEMSZ@
 disable_block_device_use = @DEFDISABLEBLOCK@
 
+# Use virtio-fs instead of 9p for shared filesystems.
+enable_virtio_fs = @DEFVIRTIOFS@
```
Will fix in the next revision.

stefanha commented on virtcontainers/device/config/config.go:
```diff
@@ -174,6 +177,9 @@ type VhostUserDeviceAttrs struct {
 	// MacAddress is only meaningful for vhost user net device
 	MacAddress string
+
+	// These are only meaningful for vhost user fs devices
```
Sorry, it's unclear from the context that additional fields are added by later patches:
        // These are only meaningful for vhost user fs devices
        Tag            string
        CacheSize      uint32
        Cache          string
        SharedVersions bool

sboeuf replied:
Got it!

stefanha commented:
I have raised https://github.com/intel/govmm/pull/91 to address the issue raised by @sboeuf.

I'm now retesting with the other virtio-fs pieces and expect to push the next revision of this pull request by the end of the week.

sboeuf commented:
@stefanha 
>I have raised intel/govmm#91 to address the issue raised by @sboeuf.
I'm now retesting with the other virtio-fs pieces and expect to push the next revision of this pull request by the end of the week.

Thanks for the update!

sboeuf commented:
@stefanha 
I have been testing the whole Kata stack last week, and I realized that from the doc, you make `virtio-fs` work with Kata by using the config option `HugePages`. This makes the guest RAM shared and backed by a file on the host, but we don't have to use huge pages. So it'd be nice to use the appropriate `FileBackedMem` option to get this to work from the `virtio-fs` documentation about how to run with Kata.
WDYT?

stefanha commented:
> @stefanha
> I have been testing the whole Kata stack last week, and I realized that from the doc, you make `virtio-fs` work with Kata by using the config option `HugePages`. This makes the guest RAM shared and backed by a file on the host, but we don't have to use huge pages. So it'd be nice to use the appropriate `FileBackedMem` option to get this to work from the `virtio-fs` documentation about how to run with Kata.
> WDYT?

Hugepages are preferred for performance but using regular MAP_SHARED pages will work.  I'll add this in a separate PR because it looks like the FileBackedMem options are just for VM templates at the moment (e.g. no memory file path is set by default) and it may need some refactoring.

stefanha commented:
I have pushed the latest code.  The Kata HowTo has been updated at https://virtio-fs.gitlab.io/howto-kata.html for anyone wishing to try this out.

sboeuf commented:
@stefanha travis CI is not passing because you need to add a field `Fixes #XXX` in one of the commit message of this PR:
```
Found 8 commits between commit 7ab69300464ab3727f7de294b564bb7224f42528 and branch master
ERROR: No "Fixes" found
ERROR: checkcommits failed. See the document below for help on formatting
commits for the project.
https://github.com/kata-containers/community/blob/master/CONTRIBUTING.md#patch-format
```

sboeuf commented:
@devimc @amshinde @ganeshmaharaj @bergwolf PTAL

devimc commented on virtcontainers/device/drivers/vhost_user_fs.go:
```diff
@@ -0,0 +1,65 @@
+// Copyright (C) 2018 Red Hat, Inc.
```
2018?

stefanha replied:
Will fix in the next revision.

devimc commented on cli/config/configuration-qemu.toml.in:
```diff
@@ -97,6 +97,32 @@ default_memory = @DEFMEMSZ@
 
+# Shared file system type:
+#   - virtio-9p (default)
+#   - virtio-fs
```
should we mention that a QEMU version greater than X is required to use this filesystem ?

stefanha replied:
It is currently not available in upstream QEMU.  Launching the guest will fail with a QEMU error message because -device vhost-user-fs-pci is not supported.  This can be seen in the kata-runtime log.

It is possible to probe a QEMU binary for this feature by running "qemu-system-x86_64 -device ?" to list available device models.  I'm not sure if that's a good idea though since it will slow down Kata.  Caching this information somewhere makes it more complex and prone to issues.  I hope just relying on the QEMU error message is sufficient.

devimc commented on virtcontainers/qemu.go:
```diff
@@ -1290,7 +1313,34 @@ func (q *qemu) addDevice(devInfo interface{}, devType deviceType) error {
+			}
+			id := hex.EncodeToString(randBytes)
+
+			sockPath, err := utils.BuildSocketPath(store.RunVMStoragePath, q.id, "vhost-fs.sock")
```
is this the same socket path used by the fs daemon?  if so, I think you should implement a new function to make sure the same path is returned in both cases.

stefanha replied:
Will fix in the next revision.

devimc commented on virtcontainers/qemu.go:
```diff
@@ -582,6 +584,27 @@ func (q *qemu) startSandbox(timeout int) error {
+		// connection with QEMU closes.  Therefore we do not keep track
+		// of this child process.
+		sourcePath := filepath.Join(kataHostSharedDir, q.id)
+		cmd = exec.Command(q.config.VirtioFSDaemon,
```
I think the daemon should start after launching qemu, the fs socket is created until QEMU is started, right? what happen if QEMU can't start? the daemon will be running forever waiting for the socket?

stefanha replied:
Will fix in the next revision.

stefanha commented:
I have pushed a new revision addressing review comments.  The biggest change is that we now wait for virtiofsd to listen on the UNIX domain socket before launching QEMU.  This avoids race conditions between the virtiofsd and QEMU processes.  There is also a cleanup function to kill virtiofsd if starting the sandbox fails.

stefanha commented:
I have addressed the "make check-go-static" failures.

stefanha commented:
/test

stefanha commented:
Can the WIP and do-not-merge labels be removed?  The WIP label is causing WIP-check to fail.

sboeuf commented:
@stefanha sorry for that! I've just removed them!

grahamwhaley commented:
Looks like it has a rebase conflict with kata-env_test.go now.
What's the status here folks - feels like we are close to a merge of virtio-fs??
/cc @ganeshmaharaj who has been looking as well I think??

stefanha commented:
> Looks like it has a rebase conflict with kata-env_test.go now.

I'm rebasing right now and will push the resolved version.

stefanha commented on virtcontainers/qemu.go:
```diff
@@ -580,13 +588,77 @@ func (q *qemu) startSandbox(timeout int) error {
+		go func() {
+			scanner := bufio.NewScanner(stderr)
+			for scanner.Scan() {
+				if strings.Contains(scanner.Text(), "Waiting for QEMU socket connection...") {
```
> This is really hacky and may block other virtiofsd implementations. One possible fix is to check the existence of the unix socket before and after `cmd.Start()`. For now, please at least add a `FIXME` above so that we can revisit it later.

Reporting readiness from a daemon is common practice.  For example, systemd has the sd_notify(3) interface where we would say "READY=1".

Polling for the existence of a UNIX domain socket has disadvantages:
1. A polling loop with a sleep() call adds latency.  The solution I've posted immediately detects that the daemon is ready without adding latency or busy loops.
2. There is a race condition between the daemon's bind(2)+listen(2) calls and QEMU's connect(2).  If QEMU is launched and attempts to connect between bind(2) and listen(2) the connect(2) call will fail.  This complicates things: either QEMU needs to add reconnect logic or kata-runtime needs to connect and pass the fd to QEMU, or the daemon needs to allow unexpected disconnects (from kata-runtime).

The only hacky thing I see here is the "QEMU" in the string we are checking.  virtiofsd works with other hypervisors and shouldn't know about QEMU.  I will change the output of virtiofsd so this string is generic.

stefanha replied:
I've submitting a virtiofsd merge request to remove references to QEMU from the output:
https://gitlab.com/virtio-fs/qemu/merge_requests/4

bergwolf replied:
@stefanha The exact virtiofsd output string becomes part of the ABI we rely on for virtiofsd readiness, -- that is the part I'm not very comfortable with. Systemd's use of `sd_notify` is standard because `READY=1` etc state change is explicitly documented in man(3) sd_notify so it becomes part of the `sd_notify` API. If we decide to use the arbitrary string as virtiofsd readiness ABI, we need to declare it loud and clear so that any virtiofsd implementation knows it **MUST** print the exactly same string to stderr to announce its readiness. Right now it is hard-coded hidden in kata runtime code... To make it more acceptable, I think at least:
1. It needs to be defined as a static variable
2. We should document it explicitly somewhere obvious to virtiofsd implementers

OTOH, it seems that other vhost user daemons in `qemu/contrib/` do not report their readiness. How do they handle the issue in practice?

bergwolf replied:
Anyway, I'm OK to merge the PR as is since the feature is experimental. We can address the readiness issue (if necessary) in follow up PRs. @stefanha please fix the other comment and we can kick off the tests to see how CI says about it.

stefanha replied:
I have pushed the latest version with GiB -> MiB fixed and QEMU -> vhost-user in the stderr message.

I will take a look at implementing sd_notify(3) in virtiofsd.  That way it will also work with systemd and we won't use stderr.  A separate file descriptor is passed for sd_notify(3) and that's a bit cleaner.  This will come as a separate pull request.

bergwolf replied:
Thanks @stefanha!

stefanha commented on cli/config/configuration-qemu.toml.in:
```diff
@@ -97,6 +97,32 @@ default_memory = @DEFMEMSZ@
+virtio_fs_daemon = "@DEFVIRTIOFSDAEMON@"
+
+# Default size of DAX cache in GiB
+virtio_fs_cache_size = @DEFVIRTIOFSCACHESIZE@
```
Yes, thank you!  Will fix in the next revision.

bergwolf commented:
/test

bergwolf commented:
The failed cases are unrelated.

sboeuf commented:
@egernst :champagne: :champagne: :champagne:
